src/Snak/ExtendedPropertyValueSnak.php

```diff
@@ -0,0 +1,87 @@
+<?php
+
+namespace Wikibase\DataModel\Snak;
+
+use DataValues\DataValue;
+use InvalidArgumentException;
+use Wikibase\DataModel\Entity\EntityId;
+use Wikibase\DataModel\Entity\PropertyId;
+
+/**
+ * @since 3.1
+ *
+ * @licence GNU GPL v2+
+ * @author Adam Shorland
+ */
+class ExtendedPropertyValueSnak extends PropertyValueSnak {
+
+	/**
+	 * @var DataValue[]
+	 */
+	private $extendedDataValues = array();
+
+	/**
+	 * Support for passing in an EntityId instance that is not a PropertyId instance has
+	 * been deprecated since 0.5.
+	 *
+	 * @param PropertyId|EntityId|integer $propertyId
+	 * @param DataValue $dataValue
+	 * @param DataValue[] $extendedDataValues
+	 */
+	public function __construct( $propertyId, DataValue $dataValue, array $extendedDataValues ) {
+		parent::__construct( $propertyId, $dataValue );
+		foreach ( $extendedDataValues as $key => $extraDataValue ) {
+			if ( !$extraDataValue instanceof DataValue ) {
+				throw new InvalidArgumentException( '$extraDataValues must be an array of DataValue objects' );
+			}
+			if ( !is_string( $key ) ) {
+				throw new InvalidArgumentException( '$extraDataValues keys must be strings' );
+			}
+		}
+		$this->extendedDataValues = $extendedDataValues;
+	}
+
+	/**
+	 * @return DataValue[]
+	 */
+	public function getExtendedDataValues() {
+		return $this->extendedDataValues;
+	}
+
+	/**
+	 * @param string $key
+	 *
+	 * @return DataValue|null
+	 */
+	public function getExtendedDataValue( $key ) {
+		if ( array_key_exists( $key, $this->extendedDataValues ) ) {
+			return $this->extendedDataValues[$key];
+		}
+		return null;
+	}
+
+	/**
+	 * @see Serializable::serialize
+	 *
+	 * @return string
+	 */
+	public function serialize() {
+		return serialize( array(
```

Comment:
There's nothing like `@override` in PHP :S Maybe add a comment that this overrides the methods defined in `PropertyValueSnak` so a) we remember to change this when `PropertyValueSnak` changes and b) nobody gets confused where there is the `implements Serializable` annotation. ;-)